src/plugins/plugin.decimation.js

```diff
@@ -0,0 +1,124 @@
+import {isNullOrUndef} from '../helpers';
+
+function minMaxDecimation(data, availableWidth) {
+	let i, point, x, y, prevX, minIndex, maxIndex, minY, maxY;
+	const decimated = [];
+
+	const xMin = data[0].x;
+	const xMax = data[data.length - 1].x;
+	const dx = xMax - xMin;
+
+	for (i = 0; i < data.length; ++i) {
+		point = data[i];
+		x = (point.x - xMin) / dx * availableWidth;
+		y = point.y;
+		const truncX = x | 0;
+
+		if (truncX === prevX) {
+			// Determine `minY` / `maxY` and `avgX` while we stay within same x-position
+			if (y < minY) {
+				minY = y;
+				minIndex = i;
+			} else if (y > maxY) {
+				maxY = y;
+				maxIndex = i;
+			}
+		} else {
+			// Push up to 4 points, 3 for the last interval and the first point for this interval
+			if (minIndex && maxIndex) {
+				decimated.push(data[minIndex], data[maxIndex]);
+			}
+			if (i > 0) {
+				// Last point in the previous interval
+				decimated.push(data[i - 1]);
+			}
+			decimated.push(point);
+			prevX = truncX;
+			minY = maxY = y;
+			minIndex = maxIndex = i;
+		}
+	}
+
+	return decimated;
+}
+
+export default {
+	id: 'decimation',
+
+	defaults: {
+		algorithm: 'min-max',
+		enabled: false,
+	},
+
+	beforeElementsUpdate: (chart, args, options) => {
+		// Decimation is only supported for lines that have an X indexAxis
+		if (!options.enabled || chart.options.indexAxis === 'y') {
+			return;
+		}
+
+		// Use an approximation since this is before scales are parsed
+		const verticalAxisCount = Object.values(chart.scales).reduce((acc, scale) => acc + scale.axis === 'y' ? 1 : 0, 0);
+
+		// assume ~50px for now. If the axis is actually wider then we have more points than needed
```

Comment:
Should we actually just assume the whole width here, because the axes can be hidden? Should still be quite efficient.
It might look bad though.

Reply:
> 
> 
> This looks like a clean solution.
> I wrote 2 comments to the available width calculation, that go in different directions regarding to the accuracy. I'm not suggesting either of those should be made here, only sharing thoughts.

Thanks @kurkle! i'll take a look and get back to you. I'm sort of leaning to the simple solution of assuming the full width since it errs on the side of adding more points.

Reply:
I changed to simplify and just assume we have `chart.width` available to use. It'll have a few extra points, but not too many.